src/datatypes/datetime.coffee

```diff
@@ -391,3 +401,81 @@ module.exports.DateTime = class DateTime
       fieldsToRemove = DateTime.FIELDS.slice(fieldIndex + 1)
       reduced[field] = null for field in fieldsToRemove
     reduced
+
+normalizeMillisecondsFieldInString = (string, matches) ->
+  msString = matches[14]
+  # TODO: verify we are only removing numeral digits
+  msString = normalizeMillisecondsField(msString)
+  [beforeMs, msAndAfter] = string.split('.')
+  timezoneSeparator = getTimezoneSeparatorFromString(msAndAfter)
+
+  timezoneField = msAndAfter?.split(timezoneSeparator)[1] if !!timezoneSeparator
+  timezoneField = '' if !timezoneField?
+  string = beforeMs + '.' + msString + timezoneSeparator + timezoneField
+
+normalizeMillisecondsField = (msString) ->
+  # fix up milliseconds by padding zeros and/or truncating (5 --> 500, 50 --> 500, 54321 --> 543, etc.)
+  msString = (msString + "00").substring(0, 3)
+
+isValidDateTimeStringFormat = (string) ->
+  return false if typeof string isnt 'string'
+  cqlFormats = ['YYYY',
+                'YYYY-MM',
+                'YYYY-MM-DD',
+                'YYYY-MM-DDZ',
+                'YYYY-MM-DDT+hh',
+                'YYYY-MM-DDT-hh',
+                'YYYY-MM-DDThh',
+                'YYYY-MM-DDThh:mm',
+                'YYYY-MM-DDThh:mmZ',
+                'YYYY-MM-DDThh:mm:ss',
+                'YYYY-MM-DDThh:mm:ssZ',
+                'YYYY-MM-DDThh:mm:ss+hh',
+                'YYYY-MM-DDThh:mm:ss+hh:mm',
+                'YYYY-MM-DDThh:mm:ss-hh',
+                'YYYY-MM-DDThh:mm:ss-hh:mm',
+                'YYYY-MM-DDThh:mm:ss.fff',
+                'YYYY-MM-DDThh:mm:ss.fffZ',
+                'YYYY-MM-DDThh:mm:ss.fff+hh',
+                'YYYY-MM-DDThh:mm:ss.fff+hh:mm',
+                'YYYY-MM-DDThh:mm:ss.fff-hh',
+                'YYYY-MM-DDThh:mm:ss.fff-hh:mm']
```

Comment:
I'm noticing (and you probably noticed too) that perhaps the CQL spec is not specific enough about what formats are allowed -- particularly in regard to where `Z` or `(+-)hh:mm` can go when the DateTime does not have complete precision.

That said, I'm not 100% sure that `YYYY-MM-DDZ` is valid.  Perhaps it should be `YYYY-MM-DDTZ` since it is a component of _time_?

In addition, it seems odd to support `YYYY-MM-DDT+hh` and `YYYY-MM-DDT-hh` but not `YYYY-MM-DDT+hh:mm` and `YYYY-MM-DDT-hh:mm`.

It also seems odd to support `YYYY-MM-DDThh:mmZ` but not the other representations of offsets for minutes.

Given that the spec is not very clear, then I'd rather err on the side of accepting too much (as long as we can parse it) than accepting too little.  For that reason, I'd suggest supporting `Z`, `+hh`, `+hh:mm`, `-hh`, and `-hh:mm` suffixes  at all levels of time precision.  Would you agree?

Reply:
> That said, I'm not 100% sure that YYYY-MM-DDZ is valid. Perhaps it should be YYYY-MM-DDTZ since it is a component of time?

That makes sense to me.

> I'd suggest supporting Z, +hh, +hh:mm, -hh, and -hh:mm suffixes at all levels of time precision. Would you agree?

Yes, that makes sense. I'll add support for those formats.